design/gateway-apis-implementation.md

```diff
@@ -1,204 +1,420 @@
 
-This pattern is applicable to all the Gateway API objects, with the exception of GatewayClass, which is not relevant for Contour.
+See: [https://github.com/projectcontour/contour/issues/3616](https://github.com/projectcontour/contour/issues/3616)
 
-There are already internal fields in the cache to hold GatewayClass and some other Gateway API objects.
-GatewayClass will be removed as part of this work, then we will use this scaffolding to build the actual functionality on top.
+### Leader Election
+
+Contour will take into account if it is the Leader or not which will represent if it should take action on managing resources or not.
+
+### Finalizers
+
+Finailzers will need to be put into place for the resources which Contour owns (i.e. Envoy) as well a GatewayAPI bits if applicable:
+
+* Envoy Service/Daemonset
+* Gateway
+* GatewayClass
+
+#### Finalizer workflows:
+
+* If the Envoy service is deleted what happens?
+* If the Envoy Daemonset is deleted what happens?
+* If the Gateway is deleted what happens?
+* If the GatewayClass is deleted what happens?
+
+### Exposing Envoy
+
+Currently, the way applications are exposed outside the cluster is done via Service Type=LoadBalancer and annotations are applied to the service causing external resources to spin up or be configured allowing external traffic to route to the Kubernetes service.
+
+Contour’s implementation of the Operator has taken a similar approach where based upon the configuration specified in the Envoy Endpoint section, the operator would translate that configuration into a set of annotations on the Envoy service.
+
+Initially Contour can do the same work to move that same logic and expose Envoy via annotations on the Service Type=LoadBalancer.
+
+The future looking way to approach this is to utilize GatewayAPI fully and instead of exposing a service with type=LoadBalancer, Contour can create a GatewayClass, Gateway which selects a TCPRoute used to send traffic to the Envoy fleet.
+The details defined in the Contour configuration CRD will be used to pre-populate annotations that will be stored in a Kubernetes configmap and referenced from the GatewayClass.Spec.ParametersRef.
+The Gateway Stub controller will be utilized to configure a new Kubernetes Service with the annotations defined in the configmap on the GatewayClass and send traffic to Envoy.
+Once implementers catch up, specific implementations will be utilized and the stub controller can be deprecated.
+
+New fields will be added to the Contour configuration file which will allow users to define the GVK of a resource used in the GatewayClass along with the ControllerName to be used.
+The Operator will still auto-create the GatewayClass/Gateway which will allow the operator to still maintain the proper listeners matching the Gateway Contour is processing, but allow for custom implementations without the need for code changes to Contour.
+
+![alt_text](images/gatewayapi-external-gateway.jpg)
+
+### WorkFlows
```

Comment:
One thing that's not described here is upgrades. Presumably the flow is - upgrade the Contour deployment, which then notices that its corresponding Envoy is out of date, and updates it? Also, do we care about downgrades / if so, do they work the same way?

If so, worth noting that this now enforces that Contour always gets upgraded before Envoy (which has been my preferred approach to upgrades, but now it's a hard requirement). That means new versions of Contour shouldn't start using new Envoy features on the critical path until they've been out in Envoy for at least a release.

Reply:
Yup I agree with what you mention @skriss. I can update the design to call that out specifically. 

I think there are also ways that we can support new Envoy features, but pick apart the Envoy version connecting. We'd need to dig into that more, but would be a neat think to at least think about for future work. I know that @sunjayBhatia had some ideas around this as well.

Reply:
Added a note about upgrades calling this out.